README.md
# ssh-connector

SSH connector which is build in from ConnId artifact


ssh-connector/CHANGELOG.md
# CHANGELOG

##### DD.MM.2020 - [1.0.1]
* .pom changes - distribution management, dependencies, maven modules





ssh-connector/src/main/java/eu/bcvsolutions/idm/connector/ssh/SSHMessages.java
/**
 * CzechIdM
 * Copyright (C) 2014 BCV solutions s.r.o., Czech Republic
 * 
 * This software is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License 2.1 as published by the Free Software Foundation;
 * 
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this software; if not, write to the Free 
 * Software Foundation, Inc., 51 Franklin Street, Fifth Floor, 
 * Boston, MA 02110-1301 USA
 * 
 * You can contact us on website http://www.bcvsolutions.eu.
 */

package eu.bcvsolutions.idm.connector.ssh;

/**
 * Deklarované konstanty.
 * 
 * @author Jaromír Mlejnek
 */
public interface SSHMessages {
	
    public static final String SSH_GETUSER = "getUser";
    public static final String SSH_CREATEUSER = "createUser";
    public static final String SSH_DELETEUSER = "deleteUser";
    public static final String SSH_ENABLEUSER = "enableUser";
    public static final String SSH_DISABLEUSER = "disableUser";
    public static final String SSH_UPDATEUSER = "updateUser";
    public static final String SSH_AUTHENTICATE = "authenticate";
    
    public static final String SSH_GETGROUP = "getGroup";
    public static final String SSH_CREATEGROUP = "createGroup";
    public static final String SSH_DELETEGROUP = "deleteGroup";
    public static final String SSH_UPDATEGROUP = "updateGroup";

    public static final String SSH_LISTOBJECTS = "listObjects";
    public static final String SSH_ATTRIBUTESLIST = "getAttributesSchema";

    public static final String SSH_HEADER_ACCOUNTID = "AccountId";
    public static final String SSH_HEADER_NEW_ACCOUNTID = "newAccountId";
    public static final String SSH_HEADER_PASSWORD = "password";
    public static final String SSH_HEADER_UID = "UID";
    public static final String SSH_HEADER_GID = "GID";
    public static final String SSH_HEADER_GECOS = "GECOS";
    public static final String SSH_HEADER_HOMEDIRECTORY = "HomeDirectory";
    public static final String SSH_HEADER_SHELL = "Shell";
    public static final String SSH_HEADER_GROUPS = "Groups";    
    public static final String SSH_HEADER_STATUS = "Status";
    
    public static final String SSH_HEADER_GROUP_NAME = "groupName";
    public static final String SSH_HEADER_USERS = "Users";
    
    public static final String SSH_OBJECT_TYPE_GROUP = "Group";
    public static final String SSH_STATUS_LOCK = "LOCK";
    public static final String SSH_STATUS_UNLOCK = "UNLOCK";
    
    public static final String SSH_HEADER_OBJECTTYPE = "objectType";
    
    public static final String SSH_ESCAPE_MODE_DOUBLED = "DOUBLED";
    public static final String SSH_ESCAPE_MODE_BACKSLASH = "BACKSLASH";
       
    public static final String SSH_ATTRIBUTE_NAME = "Attribute_Name";
    public static final String SSH_ATTRIBUTE_TYPE = "Type";
    public static final String SSH_ATTRIBUTE_FLAGS = "Flags";

}


ssh-connector/src/main/java/eu/bcvsolutions/idm/connector/ssh/SSHConnection.java
/**
 * CzechIdM
 * Copyright (C) 2014 BCV solutions s.r.o., Czech Republic
 * 
 * This software is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License 2.1 as published by the Free Software Foundation;
 * 
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this software; if not, write to the Free 
 * Software Foundation, Inc., 51 Franklin Street, Fifth Floor, 
 * Boston, MA 02110-1301 USA
 * 
 * You can contact us on website http://www.bcvsolutions.eu.
 */

package eu.bcvsolutions.idm.connector.ssh;


import java.io.IOException;

import org.identityconnectors.common.StringUtil;
import org.identityconnectors.common.logging.Log;
import org.identityconnectors.common.security.GuardedString;
import org.identityconnectors.framework.common.exceptions.ConnectionFailedException;

import com.jcraft.jsch.HostKey;
import com.jcraft.jsch.HostKeyRepository;
import com.jcraft.jsch.JSch;
import com.jcraft.jsch.JSchException;
import com.jcraft.jsch.Session;
import com.jcraft.jsch.UserInfo;

/**
 * Třída zajišťující SSH spojení s koncovým systémem.
 * 
 * @author Jaromír Mlejnek 
 */
public class SSHConnection {

	private static final Log log = Log.getLog(SSHConnection.class);	
	
	private SSHConfiguration config;
	private Session session;
	
	/**
	 * Konstruktor třídy SSHConnection.
	 * 
	 * @param cfg konfigurace, tj. instance třídy SSHConfiguration
	 * @throws Exception
	 */
	public SSHConnection(SSHConfiguration cfg) throws Exception {
		if (cfg == null) {
			throw new Exception("Configuration not set");
		}    	
		config = cfg;    	    	    	    
	} 
	
	/**
	 * Metoda pro vytvoření spojení s koncovým systémem. Pokud je uveden privátní klíč, tak
	 * se implicitně použije pro autentizaci. Jinak se použije dvojice uživatelské jméno a
	 * heslo. Pokud je uveden otisk veřejného klíče serveru, ke kterému se připojujeme, tak
	 * se použije pro jeho verifikaci.
	 *  
	 * @return Instance třídy Session. 
	 */
	public Session startConnection() {		
		String privateKey = asString(config.getPrivkey());    	    	    	
		try {    		
			if (!StringUtil.isBlank(privateKey)) {				
				//Private key used for authentication
				log.info("Private key used for authentication.");
				createSSHConnectionWithPrivateKey(privateKey, config.getPrivkeyPassword());
			} else {
				//Authentication via password
				log.info("Authentication via password.");
				createSSHConnectionWithPassword(config.getPassword());
			}
		    	
			session.connect(SSHConfiguration.CONNECTION_TIMEOUT);					
			log.info("Succesfull connection.");
		} catch (Exception ex) {
			if (ex instanceof IOException) {
				log.warn("IOException caught: " + ex.getMessage());
			} else {
				log.error("Connecting to server failed. {0}",ex.getMessage());
				throw new ConnectionFailedException("Connecting to server failed.");
			}
		} finally {
			//clear user password
			session.setPassword("");
		}
		return session;
	}
    
    /**
     * Metoda slouží pro vytvořeni SSH spojení s koncovým systémem při autentizaci 
     * uživatelským jménem a heslem.
     * 
     * @param password heslo obsažené v instanci třídy GuardedString.
     * @throws JSchException
     */
    private void createSSHConnectionWithPassword(final GuardedString password) throws JSchException {    	    	    	
    	JSch jsch = new JSch();    	
    	session = createSession(jsch);    	
    	session.setPassword(asString(password));		
    	
    	setHostHashCheckingIfNeeded(jsch, session);				    	
    }
    
    
    /**
     * Metoda slouží k vytvoření SSH spojení s koncovým systémem při autentizaci 
     * uživatelským jménem a heslem.  
     * 
     * @param privateKey privátní klíč uživatele (uzožen v jednom String řetezci).
     * @param keyPassword heslo k privátnímu klíči, pokud je uvedeno.
     * @throws JSchException
     */
    private void createSSHConnectionWithPrivateKey(String privateKey, final GuardedString keyPassword) throws JSchException {    	    	    	       	
    	privateKey = privateKey.replaceFirst("\\n(DEK-Info:[^\\n]+\\n)([^\\r\\n]+)", "$1\n$2");    	        	    	    	    	    	
    	
    	String hostkey = asString(config.getHostkey());
    	String username = config.getUsername();
    	
    	JSch jsch = new JSch();
    	
    	session = createSession(jsch);
  
    	//radek nize je tu kvuli tomu, aby se preskakovala moznost autentizace gssapi-with-mic,
    	//ktera v tomto pripade stejne nemuze uspet a zapisuje do logu vyjimku
    	session.setConfig("PreferredAuthentications","publickey,keyboard-interactive,password");
    	
    	
    	//Nacteme si pripadne heslo ke klici a otisk verejneho klice serveru.
    	boolean isKeyPasswd = (StringUtil.isBlank(asString(keyPassword))) ? false : true;
    	boolean isHostKey = (StringUtil.isBlank(hostkey)) ? false : true;    	    
    	
    	if (!isKeyPasswd) {
    		if (!isHostKey) {    			
    			log.info("Use private key - no key password, no host key.");    			
    			jsch.addIdentity(username, privateKey.getBytes(), null, null);
    		} else {
    			log.info("Use private key - no key password, specified host key.");    			
    			jsch.addIdentity(username, privateKey.getBytes(), hostkey.getBytes(), null);
    		}    	
    	}  else {
    		if (!isHostKey) {
    			log.info("Use private key - with key password, no host key.");
    			jsch.addIdentity(username, 
    					privateKey.getBytes(), null, asString(keyPassword).getBytes());
    		} else {
    			log.info("Use private key - with key password, specified host key.");
        		jsch.addIdentity(username, privateKey.getBytes(), 
        				hostkey.getBytes(), asString(keyPassword).getBytes());
    		}
    	}    	
    	
    	setHostHashCheckingIfNeeded(jsch, session);    	    	
    }
    
    /**
     * Metoda pro vytvoření relace. Je nutné, aby bylo uvedeno uživatelské jméno, 
     * adresa k serveru a číslo portu. 
     * 
     * @param jsch
     * @return Vytvořenou relaci.
     * @throws JSchException
     */
    private Session createSession(JSch jsch) throws JSchException {    	
    	Session session = jsch.getSession(config.getUsername(), 
    			config.getHost(), config.getPort());
    	return session;
    }
        
    /**
     * Metoda zajišťující kontrolu otisku veřejného klíče (pokud je zadaný).
     *  
     * @param jsch
     * @param session Relace, pro kterou se 
     */
    private void setHostHashCheckingIfNeeded(JSch jsch, Session session) {        
        String hostKeyFingerPrint = asString(config.getHostkey());    	    	
        
        if (!StringUtil.isBlank(hostKeyFingerPrint)) {        	
            HostFingerPrints fingerPrints = new HostFingerPrints(jsch);
            session.setConfig("StrictHostKeyChecking", "yes");
            fingerPrints.addFingerPrint(hostKeyFingerPrint);  
            jsch.setHostKeyRepository(fingerPrints);
        } else {
            session.setConfig("StrictHostKeyChecking", "no");
        }
    }
    
    /**
     * Metoda navracející instanci třídy SSHConfiguration (konfigurační třídy).
     * 
     * @return Instance třídy SSHConfiguration.
     */
    public SSHConfiguration getConfiguration() {
		return this.config;
	}        
	
	/**
	 * Metoda pro ukončení spojení.
	 * {@inheritDoc}
	 */
	public void dispose() {
		log.info("Dispose connection.");    	
		if (session != null) {
			session.disconnect();
		}
	}
	
	/**
	 * Metoda testující navázané spojení.
	 * {@inheritDoc}
	 */
	public void test() {
		config.validate();    	
		startConnection();
		dispose();
	}
    
    /**
     * Statická metoda, která slouží pro transformaci různých objektů na String.
     * 
     * @param object
     * @return 
     * Transformace:
     * 	GuardedString	- transformace na heslo ve Stringu.
     * 	String[]		- transformace na String, jednotlive radky pole jsou oddleny "\n".
     * 	Object 			- transformace objektu na String metodou toString.
     *  jinak			- jinak navrací null. 
     */
    public static String asString(Object object) {    	    	
        if (object instanceof GuardedString) {
            GuardedString guarded = (GuardedString)object;            
            GuardedStringAccessor accessor = new GuardedStringAccessor();
            guarded.access(accessor);
            char[] result = accessor.getArray();
            return new String(result);
        } else if (object instanceof String [] ) {
        	//Transformuje pole Stringu do jednoho Stringu.
        	StringBuilder st = new StringBuilder();
        	String item = null;
        	
            String[] array = (String[]) object;
            for (int i = 0; i < array.length; i++) {
            	item = array[i];
            	if (item == null) {
            		continue;
            	}            	
            	st.append(item);
            	st.append("\n");
            }
            return st.toString();        	        
        } else if (object != null) {
            return object.toString();
        } else {
        	return null;
        }     
    }
       
        
    static class HostFingerPrints implements HostKeyRepository {

        protected JSch jsch;
        protected String fingerPrint;

        public HostFingerPrints(JSch jsch) {
            this.jsch = jsch;
        }

        public void addFingerPrint(String fingerPrint) {
            this.fingerPrint = fingerPrint;
        }

        public int check(String arg0, byte[] arg1) {
            try {
                HostKey hostKey = new HostKey(arg0, arg1);
                String hostKeyFingerPrint = hostKey.getFingerPrint(this.jsch);
                if (hostKeyFingerPrint.equalsIgnoreCase(this.fingerPrint)) {
                    return HostKeyRepository.OK;
                } else {
                    return HostKeyRepository.CHANGED;
                }
            } catch (JSchException ex) {
                throw new RuntimeException(ex);
            }
        }	

        public void add(HostKey arg0, UserInfo arg1) {
            throw new UnsupportedOperationException("Not supported yet.");
        }

        public void remove(String arg0, String arg1) {
            throw new UnsupportedOperationException("Not supported yet.");
        }

        public void remove(String arg0, String arg1, byte[] arg2) {
            throw new UnsupportedOperationException("Not supported yet.");
        }

        public String getKnownHostsRepositoryID() {
            return null;
        }

        public HostKey[] getHostKey() {
            throw new UnsupportedOperationException("Not supported yet.");
        }

        public HostKey[] getHostKey(String arg0, String arg1) {
            throw new UnsupportedOperationException("Not supported yet.");
        }

    }
    
}


ssh-connector/src/main/java/eu/bcvsolutions/idm/connector/ssh/GuardedStringAccessor.java
/**
 * CzechIdM
 * Copyright (C) 2014 BCV solutions s.r.o., Czech Republic
 * 
 * This software is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License 2.1 as published by the Free Software Foundation;
 * 
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this software; if not, write to the Free 
 * Software Foundation, Inc., 51 Franklin Street, Fifth Floor, 
 * Boston, MA 02110-1301 USA
 * 
 * You can contact us on website http://www.bcvsolutions.eu.
 */

package eu.bcvsolutions.idm.connector.ssh;

import java.util.Arrays;

import org.identityconnectors.common.security.GuardedString;

/**
 * Třída sloužící k získání hesla z objektu třídy GuardedString.
 * 
 * @author Jaromír Mlejnek
 */
public class GuardedStringAccessor implements GuardedString.Accessor {
	
	private char[] array;
	
	/**
	 * Metoda uloži heslo z objektu třídy GuardedString do pole typu char.
	 */
	public void access(char[] clearChars) {
		array = new char [clearChars.length];
		System.arraycopy(clearChars, 0, array, 0, clearChars.length);
	}
	
	/**
	 * Metoda navracející heslo v poli typu char.
	 * 
	 * @return Pole typu char, ve kterém je uloženo heslo.
	 */
	public char[] getArray() {
		return array;
	}
	
	/**
	 * Metoda vyčistí pole, ve kterém je heslo.
	 */
	public void clearArray() {
		Arrays.fill(array, 0, array.length, ' ');
	}		

}


ssh-connector/src/main/java/eu/bcvsolutions/idm/connector/ssh/filters/SSHUserFilterTranslator.java
/**
 * CzechIdM
 * Copyright (C) 2014 BCV solutions s.r.o., Czech Republic
 * 
 * This software is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License 2.1 as published by the Free Software Foundation;
 * 
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this software; if not, write to the Free 
 * Software Foundation, Inc., 51 Franklin Street, Fifth Floor, 
 * Boston, MA 02110-1301 USA
 * 
 * You can contact us on website http://www.bcvsolutions.eu.
 */

package eu.bcvsolutions.idm.connector.ssh.filters;

import org.identityconnectors.framework.common.objects.Attribute;
import org.identityconnectors.framework.common.objects.Name;
import org.identityconnectors.framework.common.objects.Uid;
import org.identityconnectors.framework.common.objects.filter.AbstractFilterTranslator;
import org.identityconnectors.framework.common.objects.filter.EqualsFilter;
import eu.bcvsolutions.idm.connector.ssh.SSHMessages;

/**
 * Třída sloužící k vytváření vyhledávacích filtrů pro uživatelské účty.
 * 
 * @author Jaromír Mlejnek
 */
public class SSHUserFilterTranslator extends AbstractFilterTranslator<String> {	
	
	/**
	 * Metoda vytvoří dotaz (ve formě atributů pro příslušný skript koncového systému) 
	 * pro vyhledání uživatelského účtu dle specifikovaného uživatelského jména (prozatím). 
	 * Uživatelské jméno (instance třídy Name nebo Uid) je uloženo jako atribut v objektu 
	 * <b>filter</b>.	
	 */
	@Override
	protected String createEqualsExpression(EqualsFilter filter, boolean not) {		
		if (not) {
			throw new UnsupportedOperationException("Not supported yet.");
		}
		
		String username = "";
		Attribute attrib = filter.getAttribute();
		
		if (attrib == null) {
			return null;
		}
		
		if (attrib.is(Name.NAME)) {
			username = ((Name)attrib).getNameValue();
		} else if (attrib.is(Uid.NAME)) {
			username = ((Uid)attrib).getUidValue();
		} else if (attrib.getValue() != null) {
			username = (String)attrib.getValue().get(0);
		}		
		
		String operationName = SSHMessages.SSH_GETUSER;     	
		//TODO Opet pevna hodnota.
		String header = SSHMessages.SSH_HEADER_ACCOUNTID;         
	            
	    String scriptParams = String.format("%s\n%s\n%s\n", operationName, header, username);
	    
		return scriptParams;
	}
	
}


ssh-connector/src/main/java/eu/bcvsolutions/idm/connector/ssh/filters/SSHGroupFilterTranslator.java
/**
 * CzechIdM
 * Copyright (C) 2014 BCV solutions s.r.o., Czech Republic
 * 
 * This software is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License 2.1 as published by the Free Software Foundation;
 * 
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this software; if not, write to the Free 
 * Software Foundation, Inc., 51 Franklin Street, Fifth Floor, 
 * Boston, MA 02110-1301 USA
 * 
 * You can contact us on website http://www.bcvsolutions.eu.
 */

package eu.bcvsolutions.idm.connector.ssh.filters;

import org.identityconnectors.framework.common.objects.Attribute;
import org.identityconnectors.framework.common.objects.Name;
import org.identityconnectors.framework.common.objects.Uid;
import org.identityconnectors.framework.common.objects.filter.AbstractFilterTranslator;
import org.identityconnectors.framework.common.objects.filter.EqualsFilter;
import eu.bcvsolutions.idm.connector.ssh.SSHMessages;

/**
 * Třída sloužící k vytváření vyhledávacích filtrů pro uživatelské skupiny.
 * 
 * @author Jaromír Mlejnek 
 */
public class SSHGroupFilterTranslator extends AbstractFilterTranslator<String> {

	/**
	 * Metoda vytvoří dotaz (ve formě atributů pro příslušný skript koncového systému) 
	 * pro vyhledání skupiny dle specifikovaného jména skupiny (prozatím). Jméno skupiny je
	 * uloženo jako atribut v objektu <b>filter</b> jako Name nebo Uid.	
	 */
	@Override
	protected String createEqualsExpression(EqualsFilter filter, boolean not) {
		if (not) {
			throw new UnsupportedOperationException("Not supported yet.");
		}
		
		String username = "";
		Attribute attrib = filter.getAttribute();
		
		if (attrib == null) {
			return null;
		}
		
		if (attrib.is(Name.NAME)) {
			username = ((Name)attrib).getNameValue();
		} else if (attrib.is(Uid.NAME)) {
			username = ((Uid)attrib).getUidValue();
		} else {
			username = (String)attrib.getValue().get(0);
		}		
		
		String operationName = SSHMessages.SSH_GETGROUP;
		//TODO Opet pevna hodnota.
    	String header = SSHMessages.SSH_HEADER_GROUP_NAME;         
                
        String scriptParams = String.format("%s\n%s\n%s\n", operationName, header, username);
        
		return scriptParams;
	}
	
}


ssh-connector/src/main/java/eu/bcvsolutions/idm/connector/ssh/SSHConfiguration.java
/**
 * CzechIdM
 * Copyright (C) 2014 BCV solutions s.r.o., Czech Republic
 * 
 * This software is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License 2.1 as published by the Free Software Foundation;
 * 
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this software; if not, write to the Free 
 * Software Foundation, Inc., 51 Franklin Street, Fifth Floor, 
 * Boston, MA 02110-1301 USA
 * 
 * You can contact us on website http://www.bcvsolutions.eu.
 */

package eu.bcvsolutions.idm.connector.ssh;

import org.identityconnectors.common.StringUtil;
import org.identityconnectors.common.security.GuardedString;
import org.identityconnectors.framework.spi.AbstractConfiguration;
import org.identityconnectors.framework.spi.ConfigurationProperty;	

/**
 * Třída specifikující konfiguraci SSH konektoru.
 * 
 * @author Jaromír Mlejnek
 */
public class SSHConfiguration extends AbstractConfiguration {			    
       
    public static final int DEFAULT_PORT = 22;    
    public static final char DELIMITER = ';'; 
    public static final char DEFAULT_MULTIATTRIBUTE_SEPARATOR = ',';
    public static final int CONNECTION_TIMEOUT = 10000;  
    public static final String DEFAULT_EXCAPE_MODE = SSHMessages.SSH_ESCAPE_MODE_DOUBLED;
    
    private String host;
    private int port = DEFAULT_PORT;        
    private String username;
    private GuardedString password;
    
    private String[] privkey;
    private GuardedString privkeyPassword;
    private String[] hostkey;
    
    private String user;
    private String createUser;
    private String deleteUser;
    private String enableUser;
    private String disableUser;
    private String updateUser;
    private String authenticate;
    
    private String group;    
    private String createGroup;
    private String deleteGroup;
    private String updateGroup;
    
    private String listObjects;
    private String attributesSchema;
    
    private String escapeMode = DEFAULT_EXCAPE_MODE;
    private String[] multiValueAttributes;
    private char multiValueAttributesSeparator = DEFAULT_MULTIATTRIBUTE_SEPARATOR;
    
    //!!! CO SE BUDE MUSET VZDY ZADAVAT !!!
    
    @ConfigurationProperty(order = 1,
    		displayMessageKey = "SSH_HOST_NAME",
    		helpMessageKey = "SSH_HOST_HELP",
    		required = true)
    public String getHost() {
		return host;
	}
    
    @ConfigurationProperty(order = 2,
    		displayMessageKey = "SSH_PORT_NAME",
    		helpMessageKey = "SSH_PORT_HELP",
    		required = true)
    public int getPort() {
		return port;
	}
    
	@ConfigurationProperty(order = 3,
			displayMessageKey = "SSH_USER_NAME",
			helpMessageKey = "SSH_USER_HELP",
			required = true)
	public String getUsername() {
		return username;
	}
        
    @ConfigurationProperty(order = 4,
    		displayMessageKey = "SSH_USER_PASSWORD_NAME",
    		helpMessageKey = "SSH_USER_PASSWORD_HELP",
    		confidential = true)
    public GuardedString getPassword() {
		return password;
	}
    
    @ConfigurationProperty(order = 5,
    		displayMessageKey = "SSH_PRIVATE_KEY_NAME",
    		helpMessageKey = "SSH_PRIVATE_KEY_HELP")
    public String[] getPrivkey() {
		return privkey;
	}
    
    @ConfigurationProperty(order = 6,
    		displayMessageKey = "SSH_PRIVATE_KEY_PASSWORD_NAME",
    		helpMessageKey = "SSH_PRIVATE_KEY_PASSWORD_HELP",
    		confidential = true)
    public GuardedString getPrivkeyPassword() {
		return privkeyPassword;
	}
    
    @ConfigurationProperty(order = 7,
    		displayMessageKey = "SSH_HOSTKEY_NAME",
    		helpMessageKey = "SSH_HOSTKEY_HELP")
    public String[] getHostkey() {
		return hostkey;
	}
    
    //SKRIPTY
    @ConfigurationProperty(order = 8,
    		displayMessageKey = "SSH_SCRIPT_GETUSER_NAME",
    		helpMessageKey = "SSH_SCRIPT_GETUSER_HELP")
    public String getUser() {
		return user;
	}
    
    @ConfigurationProperty(order = 9,
    		displayMessageKey = "SSH_SCRIPT_CREATEUSER_NAME",
    		helpMessageKey = "SSH_SCRIPT_CREATEUSER_HELP")
    public String getCreateUser() {
		return createUser;
	}
    
    @ConfigurationProperty(order = 10,
    		displayMessageKey = "SSH_SCRIPT_DELETEUSER_NAME",
    		helpMessageKey = "SSH_SCRIPT_DELETEUSER_HELP")
    public String getDeleteUser() {
		return deleteUser;
	}
    
    @ConfigurationProperty(order = 11,
    		displayMessageKey = "SSH_SCRIPT_ENABLEUSER_NAME",
    		helpMessageKey = "SSH_SCRIPT_ENABLEUSER_HELP")
    public String getEnableUser() {
		return enableUser;
	}
    
    @ConfigurationProperty(order = 12,
    		displayMessageKey = "SSH_SCRIPT_DISABLEUSER_NAME",
    		helpMessageKey = "SSH_SCRIPT_DISABLEUSER_HELP")
    public String getDisableUser() {
		return disableUser;
	}
    
    @ConfigurationProperty(order = 13,
    		displayMessageKey = "SSH_SCRIPT_UPDATEUSER_NAME",
    		helpMessageKey = "SSH_SCRIPT_UPDATEUSER_HELP")
    public String getUpdateUser() {
		return updateUser;
	}
    
    @ConfigurationProperty(order = 14,
    		displayMessageKey = "SSH_SCRIPT_GETGROUP_NAME",
    		helpMessageKey = "SSH_SCRIPT_GETGROUP_HELP")
    public String getGroup() {
		return group;
	}
    
    @ConfigurationProperty(order = 15,
    		displayMessageKey = "SSH_SCRIPT_CREATEGROUP_NAME",
    		helpMessageKey = "SSH_SCRIPT_CREATEGROUP_HELP")
    public String getCreateGroup() {
		return createGroup;
	}
    
    @ConfigurationProperty(order = 16,
    		displayMessageKey = "SSH_SCRIPT_DELETEGROUP_NAME",
    		helpMessageKey = "SSH_SCRIPT_DELETEGROUP_HELP")
    public String getDeleteGroup() {
		return deleteGroup;
	}
    
    @ConfigurationProperty(order = 17,
    		displayMessageKey = "SSH_SCRIPT_UPDATEGROUP_NAME",
    		helpMessageKey = "SSH_SCRIPT_UPDATEGROUP_HELP")
    public String getUpdateGroup() {
		return updateGroup;
	}
    
    @ConfigurationProperty(order = 18,
    		displayMessageKey = "SSH_SCRIPT_LISTOBJECTS_NAME",
    		helpMessageKey = "SSH_SCRIPT_LISTOBJECTS_HELP")
    public String getListObjects() {
		return listObjects;
	}
    
    @ConfigurationProperty(order = 19,
    		displayMessageKey = "SSH_SCRIPT_ATTRIBUTESLIST_NAME",
    		helpMessageKey = "SSH_SCRIPT_ATTRIBUTESLIST_HELP")
    public String getAttributesSchema() {
		return attributesSchema;
	}
    
    @ConfigurationProperty(order = 20,
    		displayMessageKey = "SSH_SCRIPT_ESCAPEMODE_NAME",
    		helpMessageKey = "SSH_SCRIPT_ESCAPEMODE_HELP",
    		required = true)
    public String getEscapeMode() {
		return escapeMode;
	}
    

    @ConfigurationProperty(order = 21,
    		displayMessageKey = "SSH_SCRIPT_MULTIVALUEATT_NAME",
    		helpMessageKey = "SSH_SCRIPT_MULTIVALUEATT_HELP")
    public String[] getMultiValueAttributes() {
		return multiValueAttributes;
	}
    

    @ConfigurationProperty(order = 22,
    		displayMessageKey = "SSH_SCRIPT_MULTIVALUEATTSEPARATOR_NAME",
    		helpMessageKey = "SSH_SCRIPT_MULTIVALUEATTSEPARATOR_HELP")
    public char getMultiValueAttributesSeparator() {
		return multiValueAttributesSeparator;
	}       
    
    @ConfigurationProperty(order = 23,
    		displayMessageKey = "SSH_SCRIPT_AUTHENTICATE_NAME",
    		helpMessageKey = "SSH_SCRIPT_AUTHENTICATE_HELP")
    public String getAuthenticate() {
		return authenticate;
	}
    
    public void setHost(String host) {
		this.host = host;
	}
    
    public void setPort(int port) {
		this.port = port;
	}
    
    public void setUsername(String username) {
		this.username = username;
	}
    
    public void setPassword(GuardedString password) {
		this.password = password;
	}
    
    public void setPrivkey(String[] privkey) {
		this.privkey = privkey;
	}
    
    public void setPrivkeyPassword(GuardedString privkey_password) {
		this.privkeyPassword = privkey_password;
	}
    
    public void setHostkey(String[] hostkey) {
		this.hostkey = hostkey;
	}
    
    public void setUser(String user) {
		this.user = user;
	}
    
    public void setCreateUser(String createUser) {
		this.createUser = createUser;
	}
    
    public void setDeleteUser(String deleteUser) {
		this.deleteUser = deleteUser;
	}
    
    public void setEnableUser(String enableUser) {
		this.enableUser = enableUser;
	}
    
    public void setDisableUser(String disableUser) {
		this.disableUser = disableUser;
	}
    
    public void setUpdateUser(String updateUser) {
		this.updateUser = updateUser;
	}
    
    public void setAuthenticate(String authenticate) {
    	this.authenticate = authenticate;
    }
    
    public void setGroup(String group) {
		this.group = group;
	}
    
    public void setCreateGroup(String createGroup) {
		this.createGroup = createGroup;
	}
    
    public void setDeleteGroup(String deleteGroup) {
		this.deleteGroup = deleteGroup;
	}
    
    public void setUpdateGroup(String updateGroup) {
		this.updateGroup = updateGroup;
	}
    
    public void setListObjects(String listObjects) {
		this.listObjects = listObjects;
	}
    
    public void setAttributesSchema(String attributesSchema) {
		this.attributesSchema = attributesSchema;
	}
    
    public void setEscapeMode(String escapeMode) {
		this.escapeMode = escapeMode;
	}
    
    public void setMultiValueAttributes(String[] multiValueAttributes) {
		this.multiValueAttributes = multiValueAttributes;
	}
    
    public void setMultiValueAttributesSeparator(
			char multiValueAttributesSeparator) {
		this.multiValueAttributesSeparator = multiValueAttributesSeparator;
	}
       
    
    /**
     * Implicitní konstruktor.
     */
    public SSHConfiguration() {     	
    }
    
    /**
     * Vrací a formátuje zprávu.
     * 
     * @param key klíč pro výběr zprávy z "Message.properties".
     * @return Zpráva.
     */
    public String getMessage(String key) {
    	return getConnectorMessages().format(key, key);
    }
    
    /**
     * Vrací a formátuje zprávu.
     * 
     * @param key klíč pro výběr zprávy z "Message.properties"
     * @param objects parametry
     * @return Zpráva.
     */
    public String getMessage(String key, Object... objects) {
    	return getConnectorMessages().format(key, key, objects);
    }
    
	/**
	 * Validuje konfiguraci konektoru. Kontroluje, zda jsou nastaveny všechny potřebné parametry.
	 * Implementace by měla pouze kontrolovat syntaktickou stránku, tjn. jestli jsou vsechny potřebné
	 * parametry "well-formed". Neměla by se snažit ověřovat dostupnost zdroju, napr. pripojovat se k nim. 
	 */
	@Override
	public void validate() {    	
		if (StringUtil.isBlank(getHost())) {
			throw new IllegalArgumentException("Hostname must be set.");
		}    	
		if (getPort() < 0 || getPort() >= 65535) {
			throw new IllegalArgumentException("Port must be in range between 1 to 65535.");
		}
		if (StringUtil.isBlank(getUsername())) {
			throw new IllegalArgumentException("Username must be specified.");
		} 
		if (!getEscapeMode().equals(SSHMessages.SSH_ESCAPE_MODE_DOUBLED) && !getEscapeMode().equals(SSHMessages.SSH_ESCAPE_MODE_BACKSLASH)) {
			throw new IllegalArgumentException("Escape mode must be BACKSLASH or DOUBLED");
		}    	
	}    
    
}
